dynamiqs/result.py

```diff
@@ -287,6 +289,158 @@ class MESolveResult(SolveResult):
     """
 
 
+class MCTrajResult(SolveResult):
+    """Result of Monte Carlo trajectories.
+
+
+    Attributes:
+        states _(array of shape (..., nsave, n, 1))_: Saved states with
+            `nsave = ntsave`, or `nsave = 1` if `options.save_states` is set to `False`.
+        final_state _(array of shape (..., n, 1))_: Saved final state.
+        expects _(array of shape (..., len(exp_ops), ntsave) or None)_: Saved
+            expectation values, if specified by `exp_ops`.
+        extra _(PyTree or None)_: Extra data saved with `save_extra()` if
+            specified in `options` (see [`dq.Options`][dynamiqs.Options]).
+        infos _(PyTree or None)_: Solver-dependent information on the resolution.
+        tsave _(array of shape (ntsave,))_: Times for which results were saved.
+        solver _(Solver)_: Solver used.
+        gradient _(Gradient)_: Gradient used.
+        options _(Options)_: Options used.
+    """
+
+
+class MCSolveResult(eqx.Module):
+    """Result of Monte Carlo integration.
+
+    Attributes:
+        no_jump_states _(Array)_: Saved no-jump states.
+        final_no_jump_state _(Array)_: Saved final no-jump state.
+        jump_states _(Array)_: Saved states for jump trajectories.
+        final_jump_states _(Array)_: Saved final states for jump trajectories.
+        no_jump_prob _(Array)_: No jump probability.
+        jump_times _(Array)_: Times at which each trajectory experienced a jump. This
+            quantity has shape ..., options.max_jumps where the array is filled with
+            nans for the final options.max_jumps - num_jumps values.
+        num_jumps _(Array)_: Number of jumps each jump trajectory experienced. The times
+            at which each jump occurred is saved in jump_times.
```

Comment:
This API doesn't quite make sense to me. `mcsolve` is a function to solve the Lindblad master equation, so the only thing it must return is a set of density matrices at each time specified in `tsave`, and the corresponding `expects` is they're requested (in a very similar manner to `mesolve`). 

Of course, because it is computed in a peculiar way, we can definitely also return information regarding the no-jump/jump trajectories, but it should be secondary and IMO wrapped in a practical manner. Maybe all these should be inside a `nojump_trajectories` and `jump_trajectories` data classes, somewhat like you're implementing in the backend already?

Reply:
Thanks for this comment Ronan: I now have the states saved explicitly as density matrices as appropriate for a master-equation solver. I now also have dedicated dataclasses that save data associated with the jump and no-jump trajectories. I have made that data (states saved as kets, no-jump probability, jump times, number of jumps) accessible as part of the public api still, let me know your thoughts on that.